solution.py

```diff
@@ -0,0 +1,1188 @@
+#
+# Similar to the regular MNIST, we will use the pytorch FashionMNIST dataset. This was downloaded in the setup.sh script, so here we are just loading it into memory.
+
+# +
+fm_train_dataset = torchvision.datasets.FashionMNIST('./fashion_mnist', train=True, download=False,
+                             transform=torchvision.transforms.Compose([
+                               torchvision.transforms.ToTensor(),
+                               torchvision.transforms.Normalize(
+                                 (0.1307,), (0.3081,))
+                             ]))
+
+fm_test_dataset = torchvision.datasets.FashionMNIST('./fashion_mnist', train=False, download=False,
+                             transform=torchvision.transforms.Compose([
+                               torchvision.transforms.ToTensor(),
+                               torchvision.transforms.Normalize(
+                                 (0.1307,), (0.3081,))
+                             ]))
+# -
+
+# Next we apply the denoising model we trained on the MNIST data to FashionMNIST, and visualize the results.
+
+for i in range(8):
+    visualize_denoising(unet_model, fm_train_dataset, 123*i)
+
+# <div class="alert alert-info"><h4>
+#     Task 5.2: </h4>
+# What happened when the MNIST denoising model was applied to the FashionMNIST data? Why do you think the results look as they do?
+# </div>
+
+# + [markdown] tags=["solution"]
+# **5.2 Answer:**
+#
+# The "noise" is apparently gone, however, the objects are hardly recognizable. Some look like they have been reshaped like digits in the process.
+
+# + [markdown] tags=["solution"]
+# **5.2 Answer from 2023 Students:**
+#
+# BAD! Some of them kind of look like numbers. 
+# -
+
+# <div class="alert alert-info"><h4>
+#     Task 5.3: </h4>
+# Can you imagine any real-world scenarios where a denoising model would change the content of an image?
+# </div>
+
+# + [markdown] tags=["solution"]
+# **5.3 Answer:**
+#
+# If a denoising model is trained on data which does not appear in the data it is ultimatly used on, that new content will end up likely changed. A real worl example could be that of training a model on lots of non-dividing cells images, and use the model on new data which happens to contain some dividing cells. This could lead to the information being "denoised" away.
+
+# + [markdown] tags=["solution"]
+# **5.3 Answer from 2023**
+#
+# - Run on any out of distribution data
+# - Especially tricky if the data appears to be in distribution but has rare events. E.g. if the denoiser was trained on lots of cells that were never dividing and then was run on similar image with dividing cells, it might remove the dividing cell and replace with a single cell.
+# -
+
+# ### Train the denoiser on both MNIST and FashionMNIST
+#
+# In this section, we will perform the denoiser training once again, but this time on both MNIST and FashionMNIST datasets, and then try to apply the newly trained denoiser to a set of noisy test images.
```

Comment:
I think we need to think a bit more on what the point of this part is.  Do we actually want them to train sequentially and then shuffled? I think it would be fine to just train shuffled and show that it works and is the correct way to get good (ish?) performance on both. If we want to include a sequential example, I think it would make more sense as essentially a fine-tuning procedure, where you fully train on MNIST and then FashionMNIST, which would show how fine-tuned networks forget their initial task. The way it is implemented now, both datasets are included in both epochs so it is not really sequential or shuffled, and I find it confusing what exactly we are trying to teach them. Also, if the denoiser takes longer than 10 minutes to train, we should definitely skip sequential.

Reply:
Hey @cmalinmayor the denoiser definitely takes less to train at least in my machine with a GPU 😄 However, I am happy to skip the sequential part if you both prefer it to be removed. Let me know! :)

Reply:
I added a comment below, but long story short on the AWS hardware it is just fast enough to not be too annoying. I like having both because I like the lesson I learn from it (networks forget). I can have a go at making that a bit more on-the-nose in the text during my afternoon run typos-and-things PR :)